crates/accelerate/src/unitary_compose.rs

```diff
@@ -0,0 +1,149 @@
+        .collect::<Vec<usize>>();
+    let num_rows = usize::pow(2, num_indices as u32);
+
+    _einsum_matmul(&tensor, &mat, &indices, shift, right_mul)
+        .as_standard_layout()
+        .into_shape((num_rows, num_rows))
+        .unwrap()
+        .into_dimensionality::<ndarray::Ix2>()
+        .unwrap()
+        .to_owned()
+}
+
+// Reshape an input matrix to (2, 2, ..., 2) depending on its dimensionality
+fn per_qubit_shaped<'a>(array: &ArrayView2<'a, Complex<f64>>) -> ArrayView<'a, Complex64, IxDyn> {
+    let overall_shape = (0..array.shape()[0].ilog2() as usize)
+        .flat_map(|_| [2, 2])
+        .collect::<Vec<usize>>();
+    array.into_shape(overall_shape).unwrap()
+}
+
+// Determine einsum strings for perform a matrix multiplication on the input matrices
+fn _einsum_matmul(
+    tensor: &ArrayView<Complex64, IxDyn>,
+    mat: &ArrayView<Complex64, IxDyn>,
+    indices: &[usize],
+    shift: usize,
+    right_mul: bool,
+) -> Array<Complex64, IxDyn> {
+    let rank = tensor.ndim();
+    let rank_mat = mat.ndim();
+    if rank_mat % 2 != 0 {
+        panic!("Contracted matrix must have an even number of indices.");
+    }
+    // Get einsum indices for tensor
+    let mut indices_tensor = (0..rank).collect::<Vec<usize>>();
+    for (j, index) in indices.iter().enumerate() {
+        indices_tensor[index + shift] = rank + j;
+    }
+    // Get einsum indices for mat
+    let mat_contract = (rank..rank + indices.len()).rev().collect::<Vec<usize>>();
+    let mat_free = indices
+        .iter()
+        .rev()
+        .map(|index| index + shift)
+        .collect::<Vec<usize>>();
+    let indices_mat = if right_mul {
+        [mat_contract, mat_free].concat()
+    } else {
+        [mat_free, mat_contract].concat()
+    };
+    let tensor_einsum: String = unsafe {
+        String::from_utf8_unchecked(indices_tensor.iter().map(|c| LOWERCASE[*c]).collect())
+    };
+    let mat_einsum: String =
+        unsafe { String::from_utf8_unchecked(indices_mat.iter().map(|c| LOWERCASE[*c]).collect()) };
+    einsum(
+        format!("{},{}", tensor_einsum, mat_einsum).as_str(),
+        &[tensor, mat],
+    )
+    .unwrap()
+}
+
+fn _einsum_matmul_helper(qubits: &[u32], num_qubits: usize) -> [String; 4] {
+    let tens_in: Vec<u8> = LOWERCASE[..num_qubits].to_vec();
+    let mut tens_out: Vec<u8> = tens_in.clone();
+    let mut mat_l: Vec<u8> = Vec::with_capacity(num_qubits);
+    let mut mat_r: Vec<u8> = Vec::with_capacity(num_qubits);
+    qubits.iter().rev().enumerate().for_each(|(pos, idx)| {
+        mat_r.push(tens_in[num_qubits - 1 - pos]);
+        mat_l.push(LOWERCASE[25 - pos]);
+        tens_out[num_qubits - 1 - *idx as usize] = LOWERCASE[25 - pos];
+    });
+    unsafe {
+        [
+            String::from_utf8_unchecked(mat_l),
+            String::from_utf8_unchecked(mat_r),
+            String::from_utf8_unchecked(tens_in),
+            String::from_utf8_unchecked(tens_out),
+        ]
+    }
```

Comment:
Again, not sure if having this unsafe block here is the best way of performing this action.

Reply:
If you think it is okay and there is a good reason behind it, it would be helpful to leave a comment explaining the assumptions behind this.

Reply:
Since this was taken from my partial draft implementation that this PR was based on: https://github.com/Qiskit/qiskit/compare/main...mtreinish:qiskit-core:commutative-cancellation-no-op I'll leave a comment here about the safety. But we should add this to all the unsafe blocks in the module so we're documenting the assumptions around the safety of the blocks.

```suggestion
    // SAFETY: This is safe because we know the data in these elements is being generated solely from
    // LOWERCASE and that only contains valid utf8 characters. We don't need to spend time checking
    // if each character valid utf8 in this case.
    unsafe {
        [
            String::from_utf8_unchecked(mat_l),
            String::from_utf8_unchecked(mat_r),
            String::from_utf8_unchecked(tens_in),
            String::from_utf8_unchecked(tens_out),
        ]
    }
```

Reply:
Added a comment in 454208cd9